core/src/main/java/com/linecorp/armeria/client/endpoint/DefaultEndpointSelector.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2024 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.loadbalancer.LoadBalancer;
+import com.linecorp.armeria.common.util.ListenableAsyncCloseable;
+import com.linecorp.armeria.common.util.UnmodifiableFuture;
+
+final class DefaultEndpointSelector<T extends LoadBalancer<Endpoint, ClientRequestContext>>
+        extends AbstractEndpointSelector {
+
+    private final LoadBalancerFactory<T> loadBalancerFactory;
+    @Nullable
+    private volatile T loadBalancer;
+
+    DefaultEndpointSelector(EndpointGroup endpointGroup,
+                            LoadBalancerFactory<T> loadBalancerFactory) {
+        super(endpointGroup);
+        this.loadBalancerFactory = loadBalancerFactory;
+        if (endpointGroup instanceof ListenableAsyncCloseable) {
+            ((ListenableAsyncCloseable) endpointGroup).whenClosed().thenAccept(unused -> {
+                final T loadBalancer = this.loadBalancer;
+                if (loadBalancer != null) {
+                    loadBalancer.close();
+                }
+            });
+        }
+        initialize();
+    }
+
+    @Override
+    protected CompletableFuture<Void> updateNewEndpoints(List<Endpoint> endpoints) {
+        loadBalancer = loadBalancerFactory.newLoadBalancer(loadBalancer, endpoints);
+        return UnmodifiableFuture.completedFuture(null);
+    }
+
+    @Override
+    public Endpoint selectNow(ClientRequestContext ctx) {
+        final T loadBalancer = this.loadBalancer;
+        if (loadBalancer == null) {
+            return null;
+        }
+        return loadBalancer.pick(ctx);
```

Comment:
I understood `LoadBalancer` acts as an abstraction of picking an an `Endpoint` within an `AbstractEndpointSelector`.

i.e.

`EndpointGroup` is configured with an `EndpointSelectionStrategy`. An `EndpointSelectionStrategy` is a factory class for creating an `EndpointSelector`. An `AbstractEndpointSelector` implements asynchronous selection. `AbstractEndpointSelector` uses `LoadBalancer` to select an endpoint synchronously, or `null` if not available.